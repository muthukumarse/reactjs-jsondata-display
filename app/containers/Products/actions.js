/*
 * Products Actions
 *
*/

import {
  LOAD_PRODUCTS,
  LOAD_PRODUCTS_SUCCESS,
  LOAD_PRODUCTS_ERROR,
} from './constants';

export function loadProducts() {
  return {
    type: LOAD_PRODUCTS,
  };
}

export function productsLoaded(products) {
  return {
    type: LOAD_PRODUCTS_SUCCESS,
    payload: { products },
  };
}

export function productsLoadingError(error) {
  return {
    type: LOAD_PRODUCTS_ERROR,
    payload: { error },
  };
}
