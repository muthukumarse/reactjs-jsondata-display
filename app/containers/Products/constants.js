/*
 * Products Constants
 */

export const LOAD_PRODUCTS = 'boilerplate/Products/LOAD_PRODUCTS';
export const LOAD_PRODUCTS_SUCCESS =
  'boilerplate/Products/LOAD_PRODUCTS_SUCCESS';
export const LOAD_PRODUCTS_ERROR = 'boilerplate/Products/LOAD_PRODUCTS_ERROR';
