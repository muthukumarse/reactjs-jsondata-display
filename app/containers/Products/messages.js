/*
 * Products Messages
 */
import { defineMessages } from 'react-intl';

export const scope = 'app.containers.Products';

export default defineMessages({
  header: {
    id: `${scope}.header`,
    defaultMessage: 'All Products',
  },
  products: {
    id: `${scope}.products`,
    defaultMessage: '{totalProducts} Products',
  },
  pageSize: {
    id: `${scope}.pageSize`,
    defaultMessage: '{pageSize} per page',
  },
  previous: {
    id: `${scope}.previous`,
    defaultMessage: 'Previous',
  },
  next: {
    id: `${scope}.next`,
    defaultMessage: 'Next',
  },
});
