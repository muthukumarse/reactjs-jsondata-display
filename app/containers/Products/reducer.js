/*
 * Products Reducer
 */
import { fromJS } from 'immutable';

import {
  LOAD_PRODUCTS,
  LOAD_PRODUCTS_SUCCESS,
  LOAD_PRODUCTS_ERROR,
} from './constants';

// The initial state of the App
export const initialState = fromJS({
  products: {},
  totalProducts: 0,
  error: false,
  errorDesc: '',
});

function productsReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_PRODUCTS:
      return initialState;
    case LOAD_PRODUCTS_SUCCESS: {
      const {
        payload: { products },
      } = action;
      const totalProducts = products.length;
      return state.merge({ products, totalProducts, error: false });
    }
    case LOAD_PRODUCTS_ERROR: {
      const {
        payload: { error },
      } = action;
      return state.merge({
        products: {},
        totalProducts: 0,
        error: true,
        errorDesc: error,
      });
    }
    default:
      return state;
  }
}

export default productsReducer;
