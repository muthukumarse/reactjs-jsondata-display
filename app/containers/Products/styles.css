.products-section {
  background-color: #cbcdcf;
  bottom: 0px;
  margin: 0 auto;
  padding: 10px 20px;
  padding-top: 0;
  overflow: hidden;
}

.product-head-block {
    display: inline-block;
    width: 100%;
    border-bottom: 1px solid;
}

.product-head-block-section {
    width: 50%;
    display: inline-block;
}
.product-head-block-section-right{
    text-align: right;
}

.product-blocks {
    display: flex;
    flex-wrap: wrap;
    margin: 0 -10px;
}
.product {
    width: 300px;
    height:  320px;
    margin: 10px;
    border: 1px solid #000000;
    flex-grow: 1;
    background-color: #ffffff;
}
.product-image {
    text-align: center;
    padding: 10px 0;
    border-bottom: 1px solid #666363;
}

.product-title {
    font-weight: 400;
    padding: 10px 0 0 10px;
}

.product-desc {
    font-weight: 300;
    padding: 0 0 0 10px;
}

.product-price {
    font-weight: bold;
    padding: 0 0 0 10px;
}
.pagination-section {
    text-align: right;
}
.pagination {
    display: inline-block;
    padding-left: 0;
    margin: 20px 0;
    border-radius: 4px;
}

.pagination>li {
    display: inline;
}

.pagination>li:first-child>a, .pagination>li:first-child>span {
    margin-left: 0;
    border-bottom-left-radius: 4px;
    border-top-left-radius: 4px;
}

.pagination>li>a, .pagination>li>span {
    position: relative;
    float: left;
    padding: 6px 12px;
    line-height: 1.42857143;
    text-decoration: none;
    color: #000000;
    margin-left: -1px;
}

.pagination>li>a:hover{
  background-color: #41addd;
}

.pagination>.active>a, .pagination>.active>a:focus, .pagination>.active>a:hover, .pagination>.active>span, .pagination>.active>span:focus, .pagination>.active>span:hover {
    z-index: 3;
    color: #000000;
    background-color: #ffffff;
    border-color: #337ab7;
    border-bottom: 1px solid #000;
    cursor: default;
}

