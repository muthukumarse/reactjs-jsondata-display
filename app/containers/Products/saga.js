/**
 * Products Saga to get Data from Github
 */

import { call, put, takeLatest } from 'redux-saga/effects';
import request from 'utils/request';
import {
  productsLoaded,
  productsLoadingError,
} from 'containers/Products/actions';
import { LOAD_PRODUCTS } from './constants';

export function* getProducts() {
  const requestURL = `https://raw.githubusercontent.com/muthukumarse/reactjs-jsondata-display/master/productsData.json`;

  try {
    const products = yield call(request, requestURL);
    yield put(productsLoaded(products));
  } catch (err) {
    yield put(productsLoadingError(err));
  }
}

export default function* getProductsList() {
  yield takeLatest(LOAD_PRODUCTS, getProducts);
}
