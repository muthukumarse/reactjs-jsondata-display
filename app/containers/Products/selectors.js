/**
 * Products selectors
 */

import { createSelector } from 'reselect';
import { initialState } from './reducer';

const selectProducts = state => state.get('products', initialState);

const makeGetProducts = () =>
  createSelector(selectProducts, productsState => productsState.toJS());

export { selectProducts, makeGetProducts };
