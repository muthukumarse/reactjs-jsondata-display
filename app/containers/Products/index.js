/*
 * Products
 */

import React from 'react';
import { injectIntl, intlShape, FormattedMessage } from 'react-intl';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { createStructuredSelector } from 'reselect';
import injectReducer from 'utils/injectReducer';
import injectSaga from 'utils/injectSaga';
import Pagination from 'react-js-pagination';

import messages from './messages';
import { makeGetProducts } from './selectors';
import { loadProducts } from './actions';
import reducer from './reducer';
import saga from './saga';
import './styles.css';

export class Products extends React.Component {
  constructor(props, context) {
    super(props, context);
    this.state = {
      activePage: 1,
      pageSize: 10,
      productStartIndex: 0,
      productEndIndex: 9,
    };
  }

  componentDidMount() {
    this.props.loadProducts();
  }

  onPageChange = activePage => {
    const { pageSize } = this.state;
    const productStartIndex = pageSize * (activePage - 1);
    const productEndIndex = pageSize * activePage - 1;
    this.setState({
      activePage,
      productStartIndex,
      productEndIndex,
    });
    // console.log('activePage', activePage);
  };

  onPageSizeChange = evnt => {
    const { activePage } = this.state;
    const pageSize = evnt.target.value;
    const productStartIndex = pageSize * (activePage - 1);
    const productEndIndex = pageSize * activePage - 1;
    this.setState({
      pageSize,
      activePage: 1,
      productStartIndex,
      productEndIndex,
    });
    // console.log('pageSize', evnt.target.value);
  };

  render() {
    // console.log(this.props.productsState.products.length)
    const {
      productsState: { products, totalProducts },
      intl: { formatMessage },
    } = this.props;
    const {
      activePage,
      pageSize,
      productStartIndex,
      productEndIndex,
    } = this.state;

    if (totalProducts === 0) return <div id="loading">Loading ...</div>;

    const productList = products.slice(productStartIndex, productEndIndex + 1);
    // console.log(productStartIndex, productEndIndex);

    return (
      <div style={{ padding: '1em' }}>
        <div className="products-section">
          <h3>
            <FormattedMessage {...messages.header} />
          </h3>
          <div className="product-head-block">
            <div className="product-head-block-section">
              {formatMessage(messages.products, { totalProducts })}
            </div>
            <div className="product-head-block-section product-head-block-section-right">
              <select id="pageSize" onChange={this.onPageSizeChange}>
                <option value="10">
                  {formatMessage(messages.pageSize, { pageSize: 10 })}
                </option>
                <option value="25">
                  {formatMessage(messages.pageSize, { pageSize: 25 })}
                </option>
                <option value="50">
                  {formatMessage(messages.pageSize, { pageSize: 50 })}
                </option>
                <option value="100">
                  {formatMessage(messages.pageSize, { pageSize: 100 })}
                </option>
              </select>
            </div>
          </div>
          <div className="product-blocks">
            {productList.map(product => (
              <div className="product" key={product.id}>
                <div className="product-image">
                  <img
                    src={product.product_image}
                    alt={product.description}
                    width="200"
                    height="200"
                  />
                </div>
                <div className="product-title">{product.product_name}</div>
                <div className="product-desc">{product.description}</div>
                <div className="product-price">{product.price}</div>
              </div>
            ))}
          </div>
          <div className="pagination-section">
            <Pagination
              hideFirstLastPages
              pageRangeDisplayed={10}
              activePage={activePage}
              itemsCountPerPage={pageSize}
              totalItemsCount={totalProducts}
              onChange={this.onPageChange}
              prevPageText="Previous Page"
              nextPageText="Next Page"
            />
          </div>
        </div>
      </div>
    );
  }
}

Products.propTypes = {
  intl: intlShape,
};

export function mapDispatchToProps(dispatch) {
  return {
    loadProducts: () => dispatch(loadProducts()),
  };
}

const mapStateToProps = createStructuredSelector({
  productsState: makeGetProducts(),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

const withReducer = injectReducer({ key: 'products', reducer });
const withSaga = injectSaga({ key: 'products', saga });

export default compose(
  withReducer,
  withSaga,
  withConnect,
)(injectIntl(Products));
