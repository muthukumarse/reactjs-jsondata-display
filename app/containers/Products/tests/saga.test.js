/**
 * Tests for Products sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_PRODUCTS } from 'containers/Products/constants';
import { productsLoadingError } from 'containers/Products/actions';

import getProductsList, { getProducts } from '../saga';

describe('getProducts Saga', () => {
  let getProductsGenerator;

  beforeEach(() => {
    getProductsGenerator = getProducts();

    const selectDescriptor = getProductsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getProductsGenerator.next().value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should call the error action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getProductsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(productsLoadingError(response)));
  });
});

describe('getProductsList Saga', () => {
  const getProductsListSaga = getProductsList();

  it('should start task to watch for LOAD_PRODUCTS action', () => {
    const takeLatestDescriptor = getProductsListSaga.next().value;
    expect(takeLatestDescriptor).toEqual(
      takeLatest(LOAD_PRODUCTS, getProducts),
    );
  });
});
