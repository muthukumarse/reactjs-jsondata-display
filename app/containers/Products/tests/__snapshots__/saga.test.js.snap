// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProducts Saga should call the error action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://raw.githubusercontent.com/BookingBoss/reactjs-test/master/productsData.json",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getProducts Saga should call the error action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "products": undefined,
      },
      "type": "boilerplate/Products/LOAD_PRODUCTS_SUCCESS",
    },
    "channel": null,
  },
}
`;
